crates/sui-rest-api/src/checkpoints.rs

```diff
@@ -77,44 +80,79 @@ pub async fn get_full_checkpoint(
-        .collect::<Result<Vec<_>>>()?;
-
-    let mut transactions_effects_and_events = Vec::with_capacity(transactions.len());
+    let mut full_transactions = Vec::with_capacity(transactions.len());
     for (tx, fx) in transactions.into_iter().zip(effects) {
         let events = fx.events_digest().map(|event_digest| {
             events
                 .remove(event_digest)
                 .expect("event was already checked to be present")
         });
 
-        transactions_effects_and_events.push((tx.into(), fx, events));
+        let input_object_keys = fx
+            .input_shared_objects()
+            .into_iter()
+            .map(|(object_ref, _kind)| ObjectKey::from(object_ref))
+            .chain(
+                fx.modified_at_versions()
+                    .into_iter()
+                    .map(|(object_id, version)| ObjectKey(object_id, version)),
+            )
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        let input_objects = state
+            .database
+            .multi_get_object_by_key(&input_object_keys)?
+            .into_iter()
+            .map(|maybe_object| maybe_object.ok_or_else(|| anyhow::anyhow!("missing object")))
+            .collect::<Result<Vec<_>>>()?;
+
+        let output_object_keys = fx
+            .all_changed_objects()
+            .into_iter()
+            .map(|(object_ref, _owner, _kind)| ObjectKey::from(object_ref))
+            .collect::<Vec<_>>();
+
+        let output_objects = state
+            .database
+            .multi_get_object_by_key(&output_object_keys)?
+            .into_iter()
+            .map(|maybe_object| maybe_object.ok_or_else(|| anyhow::anyhow!("missing object")))
+            .collect::<Result<Vec<_>>>()?;
+
+        let full_transaction = CheckpointTransaction {
+            transaction: tx.into(),
+            effects: fx,
+            events,
+            input_objects,
+            output_objects,
+        };
+
+        full_transactions.push(full_transaction);
     }
 
     Ok(Bcs(CheckpointData {
         checkpoint_summary: verified_summary.into(),
         checkpoint_contents,
-        transactions: transactions_effects_and_events,
-        objects,
+        transactions: full_transactions,
     }))
 }
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct CheckpointData {
     pub checkpoint_summary: CertifiedCheckpointSummary,
     pub checkpoint_contents: CheckpointContents,
-    pub transactions: Vec<(Transaction, TransactionEffects, Option<TransactionEvents>)>,
-    pub objects: Vec<Object>,
+    pub transactions: Vec<CheckpointTransaction>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CheckpointTransaction {
+    pub transaction: Transaction,
+    pub effects: TransactionEffects,
+    pub events: Option<TransactionEvents>,
+    pub input_objects: Vec<Object>,
+    pub output_objects: Vec<Object>,
```

Comment:
do I understand correctly that input objects are the objects before execution and, if modified, their after execution would be in the output objects as well?

Reply:
correct. I'll make a comment to this effect for clarity